apollo-federation/src/sources/connect/validation/http/headers.rs

```diff
@@ -119,13 +129,82 @@ fn validate_value(
         Err(err) => return Some(err),
     };
 
-    http::HeaderValue::try_from(str_value)
+    if let Err(message) = http::HeaderValue::try_from(str_value).map_err(|_| Message {
+        code: Code::InvalidHttpHeaderValue,
+        message: format!("The value `{str_value}` at {coordinate} is an invalid HTTP header"),
+        locations: value.line_column_range(source_map).into_iter().collect(),
+    }) {
+        return Some(message);
+    }
+
+    let variable_resolver = match coordinate {
+        HttpHeadersCoordinate::Source { .. } => VariableResolver::new(
+            ConnectorsContext::new(Directive::Source, Target::RequestHeader),
+            schema,
+        ),
+        HttpHeadersCoordinate::Connect { connect, .. } => VariableResolver::new(
+            ConnectorsContext::new(connect.into(), Target::RequestHeader),
+            schema,
+        ),
+    };
+
+    // Validate that the header value can be parsed, then validate any variable expressions
+    let expression = GraphQLString::new(value, &schema.sources)
         .map_err(|_| Message {
-            code: Code::InvalidHttpHeaderValue,
-            message: format!("The value `{value}` at {coordinate} is an invalid HTTP header"),
-            locations: value.line_column_range(source_map).into_iter().collect(),
+            code: Code::GraphQLError,
+            message: format!("The value for {coordinate} must be a string."),
+            locations: value
+                .line_column_range(&schema.sources)
+                .into_iter()
+                .collect(),
         })
-        .err()
+        .ok()?;
+    match HeaderValue::from_str(str_value).map_err(|e| Message {
+        code: Code::InvalidHttpHeaderValue,
+        message: format!(
+            "The value `{str_value}` at {coordinate} is an invalid HTTP header - {message}",
+            message = e.message
+        ),
+        locations: e
+            .location
+            .and_then(|location| expression.line_col_for_subslice(location, schema))
+            .or_else(|| value.line_column_range(source_map))
+            .into_iter()
+            .collect(),
+    }) {
+        Err(message) => return Some(message),
+        Ok(header_value) => {
+            for reference in header_value.variable_references() {
+                match variable_resolver.resolve(reference, expression) {
+                    Err(message) => {
+                        return Some(Message {
+                            code: message.code,
+                            message: format!("{coordinate} contains an invalid variable reference {{{reference}}} - {message}", message = message.message),
+                            locations: message.locations,
+                        })
+                    },
+                    Ok(Some(ty)) => {
+                        if !ty.is_non_null() {
+                            return Some(Message {
+                                code: Code::NullablePathVariable,
```

Comment:
Maybe we should rename this code

Reply:
What do you mean here?

Reply:
The error code `NullablePathVariable` is not just about URL "path" anymore. So maybe like `NullabilityWarning` or something more generic that we can use?

Also, nullable URI paths might be okay in the future, depending on which proposal we go with.